content/blog/autopath.md

```diff
@@ -0,0 +1,153 @@
+domain in the search path one at a time.
+
+In this post, I'll frame the problem by covering how DNS search paths work, and the way Kubernetes
+uses search paths and how *autopath* feature can help resolve those problems.
+
+## Search Paths
+
+DNS resolvers allow for name resolution relative to their local domain. Clients can also be
+configured with a list of domains to define a search path instead of just the local domain.  When
+resolving a name, the client prefixes the name to each domain in the search path and queries the DNS
+server until a successful response is returned. If none of the queries are successful then a search
+on the absolute name is attempted.
+
+The following example shows a search for `apple`,
+
+```
+% host -ta -v apple
+Trying "apple.infoblox.com"
+Trying "apple"
+;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 25023
+;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 0
+
+;; QUESTION SECTION:
+;apple.				IN	A
+
+;; AUTHORITY SECTION:
+apple.			889	IN	SOA	a0.nic.apple. noc.afilias-nst.info. 1000002523 10800 3600 2764800 900
+```
+
+Note that the first search tried is for `apple` in the local domain `infoblox.com`. This produced no
+results, so the client then tried just plain `apple`, which resulted in an answer verifying the
+existence of the `apple.` vanity top level domain.
+
+In Unix, short name resolution behavior is more or less configured in `/etc/resolv.conf` with two
+options, `search` and `ndots` paraphrased here:
+
+- `search` Specifies the path of domains to search, in the order listed.  If not specified, the
+  search list defaults to be the local domain of the host.
+- `ndots` Sets a threshold for the number of dots which must appear in a name before an initial
+  absolute query will be made. By default, `ndots` is set to 1, which means that a query for a name
+  that contains one or more dots, such as `coredns.io` will skip the search path, and just query the
+  name absolutely i.e. `coredns.io.`.
+
+See the [resolv.conf(5)](http://man7.org/linux/man-pages/man5/resolv.conf.5.html) man page for more
+detailed descriptions.
+
+## Name Resolution in Kubernetes
+
+Kubernetes has controls the resolv.conf configuration of pods using two different DNS Policies:
+*ClusterFirst*, and *Default*.
+
+- *ClusterFirst* - Causes the pod to use a special cluster oriented search path, enabling short name resolution.
+- *Default* - Causes the pod to inherit the resolv.conf from the node it’s running on.
+
+*ClusterFirst* is the default policy (ironically). In general, with exception of some Kubernetes
+ infrastructure (notably the DNS service itself), all pods use the *ClusterFirst* policy.
+
+The *ClusterFirst* will do the following:
+- Use the cluster DNS service as the nameserver (e.g. coredns, or kube-dns)
+- Use a search path that steps "out" from the local pod's namespace.
+- Don't use the search path for searches containing 5 or more dots
+
+For example, a pod in the `default`
```

Comment:
`default` namespace:

Reply:
I would prefer `back tics` instead of *italics* for the policies, but I won't insist.